src/webauthn/proto/raw_message.rs

```diff
@@ -45,14 +58,525 @@ pub struct Packed {
+
+    fn verify_basic_constraints(&self, x509: &X509Certificate) -> Result<(), Error> {
+        if let Ok(Some(basic_constraints)) = x509.basic_constraints() {
+            if !basic_constraints.value.ca {
+                return Ok(());
+            }
+        }
+
+        Err(Error::TpmError(TpmError::CertificateRequirementNotMet(
+            "Basic Constraint".to_owned(),
+        )))
+    }
+}
+
+pub fn parse_vendor_attribute(b: &[u8]) -> IResult<&[u8], &[u8; 8]> {
+    let (b, _) = tag("id:")(b)?;
+    let (b, vendor_code) = take(8usize)(b)?;
+
+    Ok((b, vendor_code.try_into().unwrap_or(&[0u8; 8])))
+}
+
+#[derive(Default, Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct ObjectAttributes {
+    pub fixed_tpm: bool,
+    pub st_clear: bool,
+    pub fixed_parent: bool,
+    pub sensitive_data_origin: bool,
+    pub user_with_auth: bool,
+    pub admin_with_policy: bool,
+    pub no_da: bool,
+    pub encryption_duplication: bool,
+    pub restricted: bool,
+    pub decrypt: bool,
+    pub sign_or_encrypt: bool,
+}
+
+impl ObjectAttributes {
+    pub fn from_u32(o: u32) -> ObjectAttributes {
+        ObjectAttributes {
+            fixed_tpm: (o & 1) != 0,
+            st_clear: (o & 2) != 0,
+            fixed_parent: (o & 8) != 0,
+            sensitive_data_origin: (o & 16) != 0,
+            user_with_auth: (o & 32) != 0,
+            admin_with_policy: (o & 64) != 0,
+            no_da: (o & 512) != 0,
+            encryption_duplication: (o & 1024) != 0,
+            restricted: (o & 32768) != 0,
+            decrypt: (o & 65536) != 0,
+            sign_or_encrypt: (o & 131072) != 0,
+        }
+    }
+}
+
+#[derive(Default, Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct TPM2BDigest {
+    size: u16,
+    #[serde(with = "serde_bytes")]
+    buffer: Option<Vec<u8>>,
+}
+
+#[derive(Serialize, Deserialize, Default, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub enum AlgParameters {
+    #[default]
+    None,
+    RSA(RsaAlgParameters),
+    ECC(EccAlgParameters),
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct RsaAlgParameters {
+    pub symmetric: TpmAlgId,
+    pub scheme: TpmAlgId,
+    pub key_bits: u16,
+    pub exponent: u32,
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct EccAlgParameters {
+    pub symmetrics: TpmAlgId,
+    pub scheme: TpmAlgId,
+    pub curve_id: TpmEccCurve,
+    pub kdf: TpmAlgId,
+}
+
+#[derive(Default, Serialize, Deserialize, Clone, Debug)]
+#[serde(rename_all = "camelCase")]
+pub struct PublicArea {
+    #[serde(alias = "type")]
+    pub alg_type: u16,
+    pub name_alg: u16,
+    pub object_attributes: ObjectAttributes,
+    pub auth_policy: TPM2BDigest,
+    pub parameters: AlgParameters,
+    pub unique: TpmuPublicId,
+}
+
+impl PublicArea {
+    pub fn from_vec(buf: Vec<u8>) -> Result<PublicArea, Error> {
```

Comment:
```suggestion
    pub fn from_slice(buf: &[u8]) -> Result<PublicArea, Error> {
```
this works without changes and fix the issue I said in `verify_public_key`